File: bot.py
import os
import random
from telegram import Update, InlineKeyboardButton, InlineKeyboardMarkup
from telegram.ext import ApplicationBuilder, CommandHandler, CallbackQueryHandler, ContextTypes

TOKEN = os.getenv("TOKEN")

games = {}
scores = {}

WIN_COMBOS = [
    (0,1,2),(3,4,5),(6,7,8),
    (0,3,6),(1,4,7),(2,5,8),
    (0,4,8),(2,4,6)
]

def check_winner(board):
    for a,b,c in WIN_COMBOS:
        if board[a] == board[b] == board[c] != " ":
            return board[a]
    if " " not in board:
        return "Draw"
    return None

def build_board(board):
    buttons = []
    for i in range(9):
        text = board[i] if board[i] != " " else "➖"
        buttons.append(InlineKeyboardButton(text, callback_data=str(i)))
    return InlineKeyboardMarkup([buttons[i:i+3] for i in range(0,9,3)])

async def start(update: Update, context: ContextTypes.DEFAULT_TYPE):
    await update.message.reply_text(
        "🎮 *Tic Tac Toe Bot*\n\n"
        "/play – Play with a friend\n"
        "/ai – Play vs AI\n"
        "/score – Your score\n"
        "/leaderboard – Global rankings",
        parse_mode="Markdown"
    )

async def play(update: Update, context: ContextTypes.DEFAULT_TYPE):
    chat_id = update.effective_chat.id
    user = update.effective_user.first_name

    games[chat_id] = {
        "board": [" "] * 9,
        "turn": "❌",
        "names": {
            "❌": user,
            "⭕": "Opponent"
        },
        "ai": False
    }

    await update.message.reply_text(
        f"Game started!\n❌ {user}'s turn",
        reply_markup=build_board(games[chat_id]["board"])
    )

async def ai(update: Update, context: ContextTypes.DEFAULT_TYPE):
    chat_id = update.effective_chat.id
    user = update.effective_user.first_name

    games[chat_id] = {
        "board": [" "] * 9,
        "turn": "❌",
        "names": {
            "❌": user,
            "⭕": "🤖 AI"
        },
        "ai": True
    }

    await update.message.reply_text(
        f"🤖 AI Game Started!\n❌ {user}'s turn",
        reply_markup=build_board(games[chat_id]["board"])
    )

async def score(update: Update, context: ContextTypes.DEFAULT_TYPE):
    user = update.effective_user.first_name
    sc = scores.get(user, 0)
    await update.message.reply_text(
        f"🏆 *{user}*, your score: *{sc}*",
        parse_mode="Markdown"
    )

async def leaderboard(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if not scores:
        await update.message.reply_text("📉 No games played yet.")
        return

    sorted_scores = sorted(scores.items(), key=lambda x: x[1], reverse=True)

    text = "🏆 *Global Leaderboard*\n\n"
    for i, (name, sc) in enumerate(sorted_scores[:10], start=1):
        medal = "🥇" if i == 1 else "🥈" if i == 2 else "🥉" if i == 3 else "🔹"
        text += f"{medal} {i}. {name} — {sc} wins\n"

    await update.message.reply_text(text, parse_mode="Markdown")

async def move(update: Update, context: ContextTypes.DEFAULT_TYPE):
    query = update.callback_query
    await query.answer()

    chat_id = query.message.chat.id
    if chat_id not in games:
        return

    game = games[chat_id]
    board = game["board"]
    index = int(query.data)

    if board[index] != " ":
        return

    board[index] = game["turn"]
    winner = check_winner(board)

    if winner:
        if winner != "Draw":
            name = game["names"][winner]
            scores[name] = scores.get(name, 0) + 1
            text = f"🏆 {name} wins!"
        else:
            text = "🤝 It's a draw!"

        await query.edit_message_text(text, reply_markup=build_board(board))
        del games[chat_id]
        return

    game["turn"] = "⭕" if game["turn"] == "❌" else "❌"

    if game["ai"] and game["turn"] == "⭕":
        empty = [i for i, v in enumerate(board) if v == " "]
        board[random.choice(empty)] = "⭕"
        winner = check_winner(board)

        if winner:
            if winner != "Draw":
                scores["🤖 AI"] = scores.get("🤖 AI", 0) + 1
                text = "🤖 AI wins!"
            else:
                text = "🤝 It's a draw!"

            await query.edit_message_text(text, reply_markup=build_board(board))
            del games[chat_id]
            return

        game["turn"] = "❌"

    await query.edit_message_text(
        f"Turn: {game['turn']} ({game['names'][game['turn']]})",
        reply_markup=build_board(board)
    )

app = ApplicationBuilder().token(TOKEN).build()

app.add_handler(CommandHandler("start", start))
app.add_handler(CommandHandler("play", play))
app.add_handler(CommandHandler("ai", ai))
app.add_handler(CommandHandler("score", score))
app.add_handler(CommandHandler("leaderboard", leaderboard))
app.add_handler(CallbackQueryHandler(move))

print("Bot running...")
app.run_polling()

File: bot.js
require('dotenv').config();
const TelegramBot = require('node-telegram-bot-api');

const bot = new TelegramBot(process.env.BOT_TOKEN, { polling: true });

/* =========================
   SIMPLE DICTIONARY
========================= */
const dictionary = new Set([
  "apple","ant","angle","animal","axe","ball","bat","cat","car","dog","door",
  "elephant","fish","frog","goat","hat","ice","ink","jug","kite","lion","monkey",
  "nose","orange","pen","queen","rat","sun","tiger","umbrella","van","wolf","xray",
  "yak","zebra"
]);

/* =========================
   GLOBAL STATE
========================= */
const games = {};
const leaderboard = {}; // { userId: { name, wins } }

/* =========================
   HELPERS
========================= */
function uname(user) {
  return user.username ? `@${user.username}` : user.first_name;
}

function recordWin(user) {
  if (!leaderboard[user.id]) {
    leaderboard[user.id] = {
      name: uname(user),
      wins: 0
    };
  }
  leaderboard[user.id].wins++;
}

function randomLetter() {
  return String.fromCharCode(65 + Math.floor(Math.random() * 26));
}

/* =========================
   COMMANDS
========================= */
bot.onText(/\/start/, msg => {
  bot.sendMessage(
    msg.chat.id,
    `👋 Welcome ${uname(msg.from)}!\n\n` +
    `🧩 /wcg — Word Challenge Game\n` +
    `❎⭕ /xo — Tic Tac Toe\n` +
    `🏆 /leaderboard — Global rankings\n` +
    `🧑‍🤝‍🧑 join — Join lobby\n` +
    `🔄 /reset — Reset game`
  );
});

bot.onText(/\/leaderboard/, msg => {
  if (Object.keys(leaderboard).length === 0) {
    return bot.sendMessage(msg.chat.id, '📭 No games won yet.');
  }

  const sorted = Object.values(leaderboard)
    .sort((a, b) => b.wins - a.wins)
    .slice(0, 10);

  let text = '🏆 *Global Leaderboard*\n\n';
  sorted.forEach((p, i) => {
    text += `${i + 1}. ${p.name} — *${p.wins}* wins\n`;
  });

  bot.sendMessage(msg.chat.id, text, { parse_mode: 'Markdown' });
});

/* =========================
   MESSAGE HANDLER
========================= */
bot.on('message', msg => {
  const chatId = msg.chat.id;
  const text = msg.text?.trim();
  if (!text) return;

  /* RESET */
  if (text === '/reset' && games[chatId]) {
    clearTimeout(games[chatId].timer);
    delete games[chatId];
    return bot.sendMessage(chatId, '🔄 Game reset.');
  }

  /* JOIN */
  if (text.toLowerCase() === 'join' && games[chatId] && !games[chatId].started) {
    const game = games[chatId];
    if (!game.players.find(p => p.id === msg.from.id)) {
      game.players.push({
        id: msg.from.id,
        username: msg.from.username,
        first_name: msg.from.first_name
      });
      bot.sendMessage(chatId, `✅ ${uname(msg.from)} joined`);
    }
    return;
  }

  /* START WCG */
  if (text === '/wcg') {
    games[chatId] = {
      type: 'wcg',
      players: [],
      started: false,
      turn: 0,
      letter: '',
      minLen: 3,
      used: [],
      timer: null
    };

    bot.sendMessage(chatId, `🧩 *WCG Lobby*\nType *join*\n⏳ Starts in 15s`, { parse_mode: 'Markdown' });

    setTimeout(() => {
      if (games[chatId]?.players.length < 2) {
        delete games[chatId];
        return bot.sendMessage(chatId, '❌ Not enough players.');
      }
      startWCG(chatId);
    }, 15000);
    return;
  }

  /* START XO */
  if (text === '/xo') {
    games[chatId] = {
      type: 'xo',
      players: [],
      board: Array(9).fill('⬜'),
      turn: 0,
      started: false
    };

    bot.sendMessage(chatId, `❎⭕ Tic Tac Toe Lobby\nType *join*\n⏳ Starts in 15s`);

    setTimeout(() => {
      if (games[chatId]?.players.length < 2) {
        delete games[chatId];
        return bot.sendMessage(chatId, '❌ Not enough players.');
      }
      startXO(chatId);
    }, 15000);
    return;
  }

  /* WCG GAMEPLAY */
  if (games[chatId]?.type === 'wcg' && games[chatId].started) {
    const game = games[chatId];
    const player = game.players[game.turn];
    if (msg.from.id !== player.id) return;

    const word = text.toLowerCase();

    if (!word.startsWith(game.letter.toLowerCase()))
      return bot.sendMessage(chatId, '❌ Wrong starting letter');

    if (word.length < game.minLen)
      return bot.sendMessage(chatId, `❌ Minimum ${game.minLen} letters`);

    if (game.used.includes(word))
      return bot.sendMessage(chatId, '❌ Word already used');

    if (!dictionary.has(word))
      return bot.sendMessage(chatId, '❌ Not a valid word');

    game.used.push(word);
    game.minLen++;
    game.turn = (game.turn + 1) % game.players.length;
    nextWCGRound(chatId);
  }

  /* XO GAMEPLAY */
  if (games[chatId]?.type === 'xo' && games[chatId].started) {
    const game = games[chatId];
    const player = game.players[game.turn];
    if (msg.from.id !== player.id) return;

    const pos = parseInt(text) - 1;
    if (isNaN(pos) || pos < 0 || pos > 8) return;
    if (game.board[pos] !== '⬜') return;

    game.board[pos] = game.turn === 0 ? '❌' : '⭕';

    if (checkWin(game.board)) {
      recordWin(player);
      bot.sendMessage(chatId, renderBoard(game.board) + `\n🏆 ${uname(player)} wins!`);
      delete games[chatId];
      return;
    }

    game.turn = 1 - game.turn;
    bot.sendMessage(
      chatId,
      renderBoard(game.board) + `\n🎯 ${uname(game.players[game.turn])}'s turn (1–9)`
    );
  }
});

/* =========================
   WCG FUNCTIONS
========================= */
function startWCG(chatId) {
  const game = games[chatId];
  game.started = true;
  nextWCGRound(chatId);
}

function nextWCGRound(chatId) {
  const game = games[chatId];
  clearTimeout(game.timer);

  game.letter = randomLetter();
  const player = game.players[game.turn];

  bot.sendMessage(
    chatId,
    `🔤 Letter: *${game.letter}*\n👤 Player: *${uname(player)}*\n📏 Min Length: *${game.minLen}*`,
    { parse_mode: 'Markdown' }
  );

  game.timer = setTimeout(() => {
    bot.sendMessage(chatId, `⏰ ${uname(player)} eliminated`);
    game.players.splice(game.turn, 1);

    if (game.players.length === 1) {
      recordWin(game.players[0]);
      bot.sendMessage(chatId, `🏆 ${uname(game.players[0])} wins!`);
      delete games[chatId];
      return;
    }

    if (game.turn >= game.players.length) game.turn = 0;
    nextWCGRound(chatId);
  }, 20000);
}

/* =========================
   XO HELPERS
========================= */
function renderBoard(b) {
  return `${b[0]}${b[1]}${b[2]}\n${b[3]}${b[4]}${b[5]}\n${b[6]}${b[7]}${b[8]}`;
}

function checkWin(b) {
  const w = [
    [0,1,2],[3,4,5],[6,7,8],
    [0,3,6],[1,4,7],[2,5,8],
    [0,4,8],[2,4,6]
  ];
  return w.some(a => b[a[0]] === b[a[1]] && b[a[1]] === b[a[2]] && b[a[0]] !== '⬜');
}
